[AIE][Legalize] FMUL type minimization looks through fneg. FADD implemented in vector unit

martien-de-jong opened the pull request:
A widened fmul (fpext(a), fpext(b)) with possible negates is legalized to a vector MUL.
fadd(s32) is mapped to VADD.f

andcarminati commented on llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:
```diff
@@ -1674,6 +1677,11 @@ static Type *shrinkFPConstantVector(Value *V, bool PreferBFloat) {
 static Type *getMinimumFPType(Value *V, bool PreferBFloat) {
+  // Look through fneg - negation doesn't change the type requirements
+  Value *X;
+  if (match(V, m_FNeg(m_Value(X))))
```
Could we have a test?

martien-de-jong replied:
Cline complied.

martien-de-jong replied:
But Konstantin would be happier with a codegen combine.

Commit:
[AIE] Supply vector implementation for FMUL widened from S16

This is a combine pattern that pushes S32 FNEGs through FPEXT
These FNEGs prevent standard InstrCombines, and appear in some kernels.

Make buildScalarAsVector switchable by cl option to either
use broadcast or insert in undef.
The latter is easier for a follow-up combine, but is less clean because
it may have NaN effects on flags.
We wired the default to broadcast, since the combine to cover the
important case is implemented already

We also fix some cosmetics from earlier commits

andcarminati committed:
[AIE2][AIE2P] Add combiner to convert insert(undef, extract(vec, index), index) into copy

Postlegalizer combiner that matches a pattern where:
 %18:_(<16 x s32>) = COPY $x0
 %10:_(<16 x s32>) = G_IMPLICIT_DEF
 %9:_(s32) = G_CONSTANT i32 0
 %8:_(s32) = G_AIE_SEXT_EXTRACT_VECTOR_ELT %18(<16 x s32>), %9(s32)
 %22:_(<16 x s32>) = G_AIE_INSERT_VECTOR_ELT %10, %8(s32), %9(s32)

And turns it into:
 %22:_(<16 x s32>) = COPY %18

Factor out helper function

Commit:
[AIE][NFC] Make auto-update possble. The test is still the same.

andcarminati commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -16,13 +16,15 @@
 #include "AIEBaseInstrInfo.h"
 #include "AIEBaseSubtarget.h"
 #include "AIEMachineFunctionInfo.h"
+#include "MCTargetDesc/AIEMCTargetDesc.h"
```
Where is this used?

martien-de-jong replied:
Oh, vscode completion silliness.

martien-de-jong commented on llvm/test/CodeGen/AIE/GlobalISel/legalize-float-binop.mir:
```diff
@@ -4,30 +4,58 @@
+# (c) Copyright 2023-2025 Advanced Micro Devices, Inc. or its affiliates
+# RUN: llc -mtriple aie -run-pass=legalizer %s -verify-machineinstrs -o - | FileCheck --check-prefix=AIE1 %s
+# RUN: llc -mtriple aie2 -run-pass=legalizer %s -verify-machineinstrs -o - | FileCheck --check-prefix=AIE2 %s
+# RUN: llc -mtriple aie2p -run-pass=legalizer %s -verify-machineinstrs -o - | FileCheck --check-prefix=AIE2P %s
```
The file was changed by changing these lines and running the update script. I didn't check the CHECK lines, but I trust they are equivalent to the old ones.

andcarminati replied:
I prefer the new way, thanks!

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1441,19 +1443,19 @@ bool AIELegalizerHelper::legalizeG_FMUL(LegalizerHelper &Helper,
-  assert(MRI.getType(DstReg) == LLT::scalar(16) &&
-         "Expected bfloat16 type in custom legalization.");
+  assert(MRI.getType(DstReg) == S32 &&
+         "Expected float32 type in custom legalization.");
```
This used to implement a pure bf16 add/sub, by extending to f32, adding and then down converting.
We now implement the full 32 version by broadcasting S32 into the accumulators. The widening and narrowing are now done by clamping in the legalizer actions.
This has two effects:
We now have the slightly limited IEEE compliance of the vector unit. 
The down convert is now working on the extracted scalar, not directly on the vector. This lead to an extract-insert pair, which is optimized away in a separate commit, the first in this PR, thanks to Andreu

andcarminati replied:
> We now have the slightly limited IEEE compliance of the vector unit.

True, but on the other hand we already took this path for other types/ops.

andcarminati commented:
Could we also split this commit https://github.com/Xilinx/llvm-aie/pull/715/commits/e1007fa70c6698e9f7f896da50ac2594bc015fa8 in two, one for FMUL (combiner) and another one for the (FADD) legalizer?

Commit:
[AIE][Legalizer] Legalize Scalar S32 FADD/FSUB. Clamp S16 FADD/FSUB to S32

We legalize a scalar to one element in a vector. We lose some IEEE compliance,
but we get more equivalance between vectors and scalars in return.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.h:
```diff
@@ -18,6 +18,7 @@
 namespace llvm {
 
 struct AIEBaseInstrInfo;
+class MachineIRBuilder;
```
You can remove this.

andcarminati commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -23,6 +23,7 @@
 #include "llvm/CodeGen/MachineInstrBuilder.h"
 #include "llvm/CodeGen/MachineRegisterInfo.h"
 #include "llvm/CodeGen/TargetOpcodes.h"
+#include "llvm/CodeGenTypes/MachineValueType.h"
```
You can remove this.

andcarminati commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1441,19 +1442,19 @@ bool AIELegalizerHelper::legalizeG_FMUL(LegalizerHelper &Helper,
   const LLT InsertVecLLT = V16FP32;
-  SrcLHS = MIRBuilder.buildFPExt(S32, SrcLHS).getReg(0);
-  SrcRHS = MIRBuilder.buildFPExt(S32, SrcRHS).getReg(0);
+
```
I like when a great simplification can make the implementation even more general.

martien-de-jong replied:
It's the most joyful part of my job.

andcarminati commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1479,29 +1480,16 @@ bool AIELegalizerHelper::legalizeG_FADD_G_FSUB(LegalizerHelper &Helper,
 
   if (ST.isAIE2()) {
-    Res = MIRBuilder.buildBitcast(V8ACC64, Res).getReg(0);
+    if (MRI.getType(Res) != V16S32) {
```
Maybe we could use subtarget checks in a single place, to select vector operation type. Based on the selected type we can decide about `emitPadUndefVector/buildBitcast/buildUnmerge`

andcarminati replied:
like `GetShuffleModes`?

martien-de-jong replied:
Yes, I would love that, and I hope to gradually introduce more abstract concepts, like building ScalarAsVector or ScalarAsAccumulator. They would build the machine code for the smallest vector/accumulator type based on the scalar type. Just give me a month...

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3534,6 +3554,79 @@ bool llvm::matchNarrowZext(MachineInstr &MI, MachineRegisterInfo &MRI,
+
+bool llvm::matchWidenFMul(MachineInstr &FMul, MachineRegisterInfo &MRI,
+                          GISelChangeObserver &Observer, BuildFnTy &MatchInfo) {
+  if (!FMul.getMF()->getTarget().getTargetTriple().isAIE2P()) {
```
CHECK: we cover FADD/SUBB in both targets but we still may generate more libcalls for FMUL on aie2.

martien-de-jong replied:
True. I think I like to postpone that. There are a few more patterns that are missing, e.g. FABS/FNEG in vector mode etc. I think we can bundle them in a future story.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3534,6 +3554,79 @@ bool llvm::matchNarrowZext(MachineInstr &MI, MachineRegisterInfo &MRI,
+  const LLT VecTy = LLT::fixed_vector(32, S16);
+  const LLT AccTy = LLT::fixed_vector(32, S32);
+  const Register DstReg = FMul.getOperand(0).getReg();
+  bool Negate = Lft.Negate ^ Rgt.Negate;
```
nit: const Negate.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3534,6 +3554,79 @@ bool llvm::matchNarrowZext(MachineInstr &MI, MachineRegisterInfo &MRI,
+    const Register VRhs = buildScalarAsVector(B, Rgt.Source, VecTy);
+    const Register Acc = MRI.createGenericVirtualRegister(AccTy);
+    const Register Mode = B.buildConstant(S32, 60).getReg(0);
+    B.buildIntrinsic(Intrinsic::aie2p_I512_I512_ACC1024_bf_mul_conf, Acc, true,
```
Could we move the instrinsic selection to the beginning? Based on the target? In case we don't have an intrinsic we can skip this combiner. It may not solve directly the problem of porting this combiner to other targets, but at least it can help by making it more clear (types are also target independent).

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3534,6 +3554,79 @@ bool llvm::matchNarrowZext(MachineInstr &MI, MachineRegisterInfo &MRI,
+      LSrc = MRI.createGenericVirtualRegister(S16);
+      B.buildInstr(TargetOpcode::G_FNEG, {LSrc}, {Lft.Source});
+    }
+    const Register VLhs = buildScalarAsVector(B, LSrc, VecTy);
```
Nice!

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3534,6 +3554,79 @@ bool llvm::matchNarrowZext(MachineInstr &MI, MachineRegisterInfo &MRI,
+    const Register VLhs = buildScalarAsVector(B, LSrc, VecTy);
+    const Register VRhs = buildScalarAsVector(B, Rgt.Source, VecTy);
+    const Register Acc = MRI.createGenericVirtualRegister(AccTy);
+    const Register Mode = B.buildConstant(S32, 60).getReg(0);
```
Maybe this mode could also be moved up (just the value), together with the intrinsic ID. I think we should make it visible as a target-specific mode.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -3534,6 +3554,79 @@ bool llvm::matchNarrowZext(MachineInstr &MI, MachineRegisterInfo &MRI,
+        .addUse(VRhs)
+        .addUse(Mode);
+    const Register Index = B.buildConstant(S32, 0).getReg(0);
+    B.buildInstr(TII.getGenericExtractVectorEltOpcode(/*SignExt*/ true),
```
This SignExt is always funny for 32bit extractions.

andcarminati replied:
Could we also have something that reverses `buildScalarAsVector`, like `convertVectorToScalar`?

andcarminati replied:
We can have an contract related to element zero.

martien-de-jong replied:
Even funnier for float types...

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/prelegalizercombiner-widen-fmul16.mir:
```diff
@@ -0,0 +1,148 @@
+    $r0 = COPY %8(s32)
+    PseudoRET implicit $lr, implicit $r0
+...
+
```
Too much lines here!

martien-de-jong replied:
No, too many!

andcarminati committed:
[AIEX][GlobalISel] Add broadcast-extract-to-copy combiner for AIE2/AIE2P

This patch adds a new postlegalizer combiner that optimizes patterns where
an element is extracted from position 0, broadcast to a vector, and then
only position 0 is used through a chain of operations.

Pattern optimized:
  %extract = G_AIE_SEXT_EXTRACT_VECTOR_ELT %src(<N x T>), 0
  %broadcast = G_AIE_BROADCAST_VECTOR %extract
  ... (chain of concat/unmerge/vector ops)
  %result = G_AIE_SEXT_EXTRACT_VECTOR_ELT %final, 0

Transforms to:
  %extract = G_AIE_SEXT_EXTRACT_VECTOR_ELT %src(<N x T>), 0
  %broadcast = COPY %src  // Broadcast eliminated!
  ... (chain of operations)
  %result = G_AIE_SEXT_EXTRACT_VECTOR_ELT %final, 0

The combiner uses a conservative whitelist approach to verify that operations
in the chain don't shift vector elements. It handles:
- G_CONCAT_VECTORS (broadcast must be first operand)
- G_UNMERGE_VALUES (only first output used, others dead)
- Whitelisted operations: G_FADD, G_FMUL, G_FSUB, and specific AIE2P intrinsics
- Rejects: G_BITCAST, type mismatches, non-zero extracts, multiple uses

This optimization is particularly beneficial when broadcasts are used unnecessarily when
only the first element is needed.

The combiner runs as a postlegalizer custom combiner for both AIE2 and AIE2P.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -68,6 +73,15 @@ cl::opt<bool> MemsetOptimizations(
 
 namespace {
 
+static constexpr const LLT S8 = LLT::scalar(8);
```
Do we need static here?

martien-de-jong replied:
No, nor constexpr...

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -123,40 +137,107 @@ bool verifyBroadcastUsesOnlyExtractZero(Register Reg, MachineRegisterInfo &MRI,
 }
 
-} // namespace
+Register buildInsertInUndef(MachineIRBuilder &B, Register Src, LLT VecTy) {
```
We are kind of generating legal code here. For example, we use AIE-specific G opcode that demands s32 types. On the other hand, we could also use standard generic opcods that can work on vector Elt type to insert - and let the legalizer do the rest.

I recommend to consider this for a follow up PR - not now.

martien-de-jong replied:
https://jira.xilinx.com/browse/AIECC-1092

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -123,40 +137,107 @@ bool verifyBroadcastUsesOnlyExtractZero(Register Reg, MachineRegisterInfo &MRI,
+  const Register IdxReg = B.buildConstant(S32, 0).getReg(0);
+  const Register UndefVec = B.buildUndef(VecTy).getReg(0);
+  const unsigned InsertEltOpc = TII.getGenericInsertVectorEltOpcode();
+  Register Vector =
```
nit: const.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -123,40 +137,107 @@ bool verifyBroadcastUsesOnlyExtractZero(Register Reg, MachineRegisterInfo &MRI,
+  }
+  const AIEBaseInstrInfo &TII = getAIETII(B);
+  const unsigned InsertEltOpc = TII.getGenericBroadcastVectorOpcode();
+  Register Vector = B.buildInstr(InsertEltOpc, {VecTy}, {Src}).getReg(0);
```
nit: const.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -123,40 +137,107 @@ bool verifyBroadcastUsesOnlyExtractZero(Register Reg, MachineRegisterInfo &MRI,
+  return Vector;
+}
+
+Register buildScalarAsVector(MachineIRBuilder &B, Register Src, LLT VecTy) {
```
If we take this a common utility, we could refactor also legalization code (future).

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -123,40 +137,107 @@ bool verifyBroadcastUsesOnlyExtractZero(Register Reg, MachineRegisterInfo &MRI,
+  return Acc;
+}
+
+void buildFirstElement(MachineIRBuilder &B, Register DstReg, Register Vec) {
```
Also a good candidate for utility function. We could also have something like:

andcarminati replied:
What about `Register buildGetFirstElement` ?

andcarminati reviewed (approved):
LGTM.
